File: Server/Extensions/Marker/Marker.cs
using System.Drawing;
using System.Numerics;
using AltV.Net.Data;

namespace Server.Extensions.Marker
{
    public class Marker
    {
        public int Type { get; set; }
        public float PosX { get; set; }
        public float PosY { get; set; }
        public float PosZ { get; set; }
        public float DirX { get; set; }
        public float DirY { get; set; }
        public float DirZ { get; set; }
        public float RotX { get; set; }
        public float RotY { get; set; }
        public float RotZ { get; set; }
        public float Scale { get; set; }
        public Color Color { get; set; }
        public float Range { get; set; }
        public bool Bob { get; set; }
        public bool FaceCamera { get; set; }
        public bool Rotate { get; set; }

        /// <summary>
        /// Returns a New Marker Object
        /// </summary>
        /// <param name="type"></param>
        /// <param name="position"></param>
        /// <param name="direction"></param>
        /// <param name="rotation"></param>
        /// <param name="scale"></param>
        /// <param name="color"></param>
        /// <param name="range"></param>
        /// <param name="rotate"></param>
        /// <param name="bob"></param>
        /// <param name="faceCamera"></param>
        public Marker(MarkerType type, Position position, Vector3 direction, Rotation rotation, float scale, Color color, float range = 5f,
            bool rotate = false, bool bob = false, bool faceCamera = false)
        {
            Type = (int)type;
            PosX = position.X;
            PosY = position.Y;
            PosZ = position.Z;
            DirX = direction.X;
            DirY = direction.Y;
            DirZ = direction.Z;
            RotX = rotation.Pitch;
            RotY = rotation.Roll;
            RotZ = rotation.Yaw;
            Scale = scale;
            Color = color;
            Range = range;
            Rotate = rotate;
            Bob = bob;
            FaceCamera = faceCamera;
        }
    }
}

File: Client/src/marker.ts
import * as alt from 'alt-client';
import * as native from 'natives';
import * as extensions from './extensions';

var markerJsonArray = new Array();
var markerArray = new Array();
alt.onServer('createMarker', createNewMarker);
alt.onServer('deleteMarker', deleteMarker);

alt.onServer('deleteAllMarkers', () => {
    markerJsonArray = new Array();
    markerArray = new Array();
});
function createNewMarker(markerJson: string) {
    markerJsonArray.push(markerJson);
    var marker = JSON.parse(markerJson);
    marker.pos = new alt.Vector3(marker.PosX, marker.PosY, marker.PosZ);
    markerArray.push(marker);
}
function displayMarker(marker: any) {
    let playerPos = alt.Player.local.pos;
    let playerDist = extensions.Distance(playerPos, marker.pos);
    if (playerDist < marker.Range) {
        native.drawMarker(marker.Type, marker.PosX, marker.PosY, marker.PosZ, marker.DirX, marker.DirY, marker.DirZ, marker.RotX, marker.RotY, marker.RotZ, marker.Scale, marker.Scale, marker.Scale, marker.Color.R, marker.Color.G, marker.Color.B, marker.Color.A, marker.Bob, marker.FaceCamera, 2, marker.Rotate, undefined, undefined, false);
    }
}
function deleteMarker(markerJson: string) {
    for (let index = 0; index < markerJsonArray.length; index++) {
        const element = markerJsonArray[index];
        if (element === markerJson) {
            markerJsonArray.splice(index, 1);
        }
    }
    for (let index = 0; index < markerArray.length; index++) {
        const element = markerArray[index];
        if (element === JSON.parse(markerJson)) {
            markerArray.splice(index, 1);
        }
    }
}
function processMarkers() {
    if (markerArray.length == 0)
        return;
    markerArray.forEach(element => {
        displayMarker(element);
    });
}
alt.setInterval(() => {
    processMarkers();
}, 0);
